keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,724 @@
+1. `bucketAccessClassName`: (required) name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance. Set by COSI.
+1. `accessGranted`: if true access has been granted to the bucket. If false then access to the bucket has not been granted.
+
+#### BucketAccess
+
+A cluster-scoped administrative custom resource which encapsulates fields from the `BucketAccessRequest` (BAR) and the `BucketAccessClass` (BAC).  The purpose of the `BucketAccess` (BA) is to serve as an access communication path between provisioners and the central COSI controller.  The COSI controller creates a `BucketAccess` instance for a new `BucketAccessRequest`, and there is a 1:1 mapping between `BucketAccess` and `BucketAccessRequest`.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccess
+metadata: 
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+ spec:
+  provisioner: [4]
+  bucketInstanceName: [5]
+  bucketAccessRequest: [6]
+  serviceAccount: [7]
+  mintedSecretName: [8]
+  policyActionsConfigMapData: [9]
+  principal: [10]
+  parameters: [11]
+ status:
+   accessGranted: [12]
+```
+
+1. `name`: when created by COSI, the `BucketAccess` name is generated in this format: _"ba-"+uuid_. If an admin creates a `BucketAccess`, as is necessary for driverless access, they can use any name. The uuid is unique within a cluster.
+1. `labels`: added COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by COSI to defer `BucketAccess` deletion until the pod that uses this access terminates, and the related `BucketAccessRequest` is deleted.
+1. `provisioner`: (optional) The provisioner field as defined in the `BucketAccessClass`. If empty then there is no driver/sidecar, thus the admin had to create the `Bucket` and `BucketAccess` instances.
+1. `bucketInstanceName`:  name of the `Bucket` instance bound to this BA.
+1. `bucketAccessRequest`: an `objectReference` containing the name, namespace and UID of the associated `BucketAccessRequest`.
+1. `serviceAccount`: an `ObjectReference` containing the name, namespace and UID of the associated `BAR.serviceAccountName`. If empty then integrated Kubernetes -> cloud identity is not being used, in which case, `BucketAccess.principal` contains the user identity, which is minted by the provisioner.
+1. `mintedSecretName`: name of the provisioner-generated Secret containing access credentials. This Secret exists in the provisioner’s namespace and is copied to the app namespace by the COSI controller.
```

Comment:
>> Can you describe this process? Who creates mintedSecretName in provisioner's namespace? How do you prevent name collisions? Why does it need to be copied to the app namespace? Why can't secret just be provisioned in app namespace?
> 
> @jeffvance jeffvance 7 days ago Author
> A provisioner does not have access to the user's namespace, and the minted secret lives in provisioner namespace. This secret is created by either the admin (static) or by COSI when it is the result of backend bucket access creation. It contains the credential returned by the driver. Name collisions are prevented by using UUID(secret-uuid).

Why does it need to be copied to the app namespace? Why can't the secret just be provisioned in app namespace to begin with?

Reply:
Copying the secret is a carryover from an earlier design. Will update kep to reflect the current design, which adheres to _minimal privileges_ such that the provisioner does not need permissions to access user namespaces. Instead, the cosi-node-adapter will write the secret contents (provisioner's ns) to the app pod's secret mount.